File: app/src/main/java/com/solution/it/newsoft/paging/ListDataFactory.java
package com.solution.it.newsoft.paging;

import com.solution.it.newsoft.datasource.Repository;

import javax.inject.Inject;

import androidx.lifecycle.MutableLiveData;
import androidx.paging.DataSource;

public class ListDataFactory extends DataSource.Factory {

    private MutableLiveData<ListDataSource> mutableLiveData;
    private ListDataSource feedDataSource;
    private Repository repository;

    @Inject
    public ListDataFactory(Repository repository) {
        this.repository = repository;
        this.mutableLiveData = new MutableLiveData<>();
    }

    @Override
    public ListDataSource create() {
        feedDataSource = new ListDataSource(repository);
        mutableLiveData.postValue(feedDataSource);
        return feedDataSource;
    }

    public void reload() {
        if (feedDataSource != null)
            feedDataSource.invalidate();
    }


    public MutableLiveData<ListDataSource> getMutableLiveData() {
        return mutableLiveData;
    }
}

File: app/src/main/java/com/solution/it/newsoft/paging/ListDataFactory.kt
package com.solution.it.newsoft.paging

import com.solution.it.newsoft.datasource.Repository

import androidx.lifecycle.MutableLiveData
import androidx.paging.DataSource
import com.solution.it.newsoft.model.List
import javax.inject.Inject

class ListDataFactory @Inject constructor(private val repository: Repository) : DataSource.Factory<Long, List>() {

    val mutableLiveData: MutableLiveData<ListDataSource> = MutableLiveData()
    private lateinit var feedDataSource: ListDataSource

    override fun create(): ListDataSource {
        feedDataSource = ListDataSource(repository)
        mutableLiveData.postValue(feedDataSource)
        return feedDataSource
    }

    fun reload() = feedDataSource.invalidate()

    fun retry() = feedDataSource.retry()
}
